Add export-enrolled-keys command

jimmykarily opened the pull request:
This PR introduces an `export-enrolled-keys` command that allows the user to export the already enrolled keys to a directory for backup reasons (e.g. before making any changes on new hardware) and to prepare a directory that can be used as a "custom certs" when enrolling keys.

User scenario:

- User buys some new hardware with MS and vendor keys already enrolled
- User backs up the built-in keys using this new command
- User generates new keys (intended to sign their own OSes)
- User puts the machine in "setup mode" by deleting PK, KEK and db from bios
- User enrolls the new keys including the original ones as "custom keys"

This PR is only one of the things needed to implement the above scenario. Currently the export directory structure is the one expected by the `enroll-keys` command but the location is arbitrary. The user would have to copy the directory to the appropriate location (the db directory) in order to be read by the "enroll-keys" command.

In my opinion, it's the `enroll-keys` command that should be changed to be more flexible and read keys from arbitrary locations but this can be discussed on another ticket/PR.

Note: I would have written a test for this but waiting for this refactoring first: https://github.com/Foxboron/sbctl/pull/302

I'm opening this draft PR to collect some early feedback, until I can write the tests.

Foxboron commented:
We already have `sbctl enroll-keys --export` which exports the keys `sbctl` intends to enroll into esl or auth files. I'm wondering if we should try to mold this into something you can use instead of adding another command.

Would `sbctl enroll-keys --export der` be a thing you could use?

jimmykarily commented:
> We already have `sbctl enroll-keys --export` which exports the keys `sbctl` intends to enroll into esl or auth files. I'm wondering if we should try to mold this into something you can use instead of adding another command.
> 
> Would `sbctl enroll-keys --export der` be a thing you could use?

I may be missing something but the use case of the new command is to export the build-in certificates from brand new hardware. I thought that maybe a combination of flags of the `enroll-keys` could achieve the same thing but it doesn't seem to be the case:

```
~/workspace/sbctl (export-enrolled-keys)*$ sudo go run ./cmd/sbctl enroll-keys --firmware-builtin --export esl
Found OptionROM in the bootchain. This means we should not enroll keys into UEFI without some precautions.

There are three flags that can be used:
    --microsoft: Enroll the Microsoft OEM certificates into the signature database.
    --tpm-eventlog: Enroll OpRom checksums into the signature database (experimental!).
    --yes-this-might-brick-my-machine: Ignore this warning and continue regardless.

Please read the FAQ for more information: https://github.com/Foxboron/sbctl/wiki/FAQ#option-rom
exit status 1
``` 

I could try setting `--yes-this-might-brick-my-machine` but even if this worked, it still feels wrong. Exporting keys is a non mutating action and shouldn't print any warnings at all (or required "dangerous" flags).

#### Kairos specific background for better understanding

To make it even more clear, this is our full user scenario:

- User maintains their own PKI infrastructure. This means they generate private and public keys by different means (not necessarily sbctl).
- User wants to sign Kairos OS with their own keys but we also want the full bundle of keys (the user's + the original ones from the hardware) to be enrolled automatically on first boot in setup mode (systemd-boot feature).

The above is an one-off procedure. As soon as the Kairos OS has been prepared, any machine booting that OS in setup mode, will automatically get the same certificates enrolled. This allows the user to prepare thousands of machines with less effort.

We already have [tooling that will prepare the OS](https://github.com/kairos-io/enki/blob/dc860f497152ec93dd025e6345d7c957ea6525b3/cmd/build-uki.go#L19) and the input to that is the set of keys to the efi file but also the keys to enroll automatically when the system boots.

So, what we need from sbctl is to allow the user to prepare such a set of keys in an easy way. The process I had in mind was:

- use sbctl to export the original keys
- use sbctl to combine user supplied keys with the above exported keys into authenticated variable files that can be enrolled with systemd-boot

the result of the above will be fed to the `enki build-uki` command (linked above) to prepare the OS.

jimmykarily commented:
It seems that sbctl can do all things already for us except maybe exporting the builtin keys to be used later on another machine to prepare the rest of the keys. We are investigating here: https://github.com/kairos-io/enki/issues/91

@nianyush will give it a try with `--yes-this-might-brick-my-machine` in a VM and maybe if that works, we can just improve the wording or skip the need for the flag when the `--export` flags is also set.

Foxboron commented:
I suspect the code paths is a bit non-obvious which is why the optionrom check is being triggered. I think we should be able to move things around for this to work properly :)

If this doesn't work at all or becomes too confusing I'm totes open for the original idea of having a new command.

nianyush commented:
i can confirm that these two commands works for me to generate the auth file.
```
sbctl enroll-keys --export auth --yes-this-might-brick-my-machine -a
sbctl enroll-keys --export auth --yes-this-might-brick-my-machine -f
```
Although in qemu vm, -f option didnt work for me. It seems my qemu vm is missing dbDefault efivar. However, i can still achieve my goal with -a since i have my factory keys already enrolled in the system.
```
root@ubuntu:/home/nianyu/work# sbctl enroll-keys --export auth --yes-this-might-brick-my-machine -f
Exporting keys to EFI files...
With vendor certificates built into the firmware...✗ 
couldn't sync keys: could not enroll built-in firmware keys: open /sys/firmware/efi/efivars/dbDefault-8be4df61-93ca-11d2-aa0d-00e098032b8c: no such file or directory
root@ubuntu:/sys/firmware/efi/efivars# ls | grep db
certdb-d9bee56e-75dc-49d9-b4d7-b534210f637a
certdbv-d9bee56e-75dc-49d9-b4d7-b534210f637a
db-d719b2cb-3d3a-4596-a3bc-dad00e67656f
dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f
``` 
@jimmykarily @Foxboron

Foxboron commented:
Cool, so we can fix the UX here to make it better and easier to work with then introducing a new command :)

nianyush commented:
Yes, and one more command i think it might be helpful is to export auth/esl from enrolled keys in the system directly. Because currently `enroll-keys -a --export auth` would need me to first create some keys and then merge my created keys with system keys.

nianyush commented:
it will be similar to
```
efi-readvar -v db -o 'db.esl'
```

jimmykarily commented:
> Yes, and one more command i think it might be helpful is to export auth/esl from enrolled keys in the system directly.

You mean like the command introduced in this PR? I'm confused now :D

nianyush commented:
@jimmykarily --export works on some degree if you have your custom keys imported using sbctl import-keys. But if you dont have keys created/imported by sbctl. It would fail
```
root@ubuntu:/usr/share/secureboot# sbctl enroll-keys --export auth  --yes-this-might-brick-my-machine
open /usr/share/secureboot/GUID: no such file or directory
``

jimmykarily commented:
@nianyush so the desired command would simply export the enrolled keys without importing anything first. If you check this PR code, that's exactly what it does (exports the "der" version but that we can change or make configurable). So what I'm trying to understand is whether the command introduced in this PR is still needed or not.

nianyush commented:
Then the question would probably be yes for me:) @jimmykarily

jimmykarily commented:
I've added a couple of tests and added a flag for the export format ("der" or "esl").

@Foxboron if you prefer this functionality to be implemented by the `enroll-keys` with additional flags let me know and I can move it there. In my opinion, one command should do one thing. `enroll-keys` should enroll, `export-whatever` should export. And there is the distinction between "export enrolled keys to the database" and "export the enrolled key to the specified directory". I would like to keep the option to export to a specified directory because the database path may not be writable (e.g. in Kairos).

Foxboron commented:
> if you prefer this functionality to be implemented by the enroll-keys with additional flags let me know and I can move it there. In my opinion, one command should do one thing. enroll-keys should enroll, export-whatever should export. 

I agree that lumping this up with `enroll-keys` would be confusing. So we can have this command while we figure out a better approach to exporting the auth lists.

jimmykarily commented on tests/integration_test.go:
```diff
@@ -27,7 +27,13 @@ func (vm *VMTest) RunTests(packages ...string) func(t *testing.T) {
 			vmtest.WithVMOpt(
 				vmtest.WithSharedDir("ovmf/keys"),
-				vmtest.WithInitramfsFiles("sbctl:bin/sbctl"),
+				vmtest.WithInitramfsFiles(
+					"sbctl:bin/sbctl",
+					"/usr/bin/openssl:bin/openssl",
+					"/usr/bin/find:bin/find",
+					"/usr/bin/head:bin/head",
+					"/usr/bin/sig-list-to-certs:bin/sig-list-to-certs",
+				),
```
You mean, implement them in Go, build a single binary and inject that helper binary in the initramfs here (instead of the individual ones)?

That would make tests less dependent on the host OS, which is good. I just want to make sure I understand your suggestion correctly.

Foxboron replied:
Ah no, I mean it's simpler to just implement these things as helper functions instead of injecting binaries into the `initramfs`.

The helper functions can also be shared between tests which is going to be nicer as well.

jimmykarily replied:
Hm, that would need me to somehow get the resulting files out of the VM to inspect them in "go space". I don't see a way in the `vmtest` library to either copy files from the VM to the host or share a directory somehow. Maybe this? https://github.com/hugelgupf/vmtest/blob/5d9f3d34a58dc7b13bca786e8ac32d3e2ce4e95d/examples/shareddir/vm_test.go#L30

Foxboron replied:
Ah, then you are misunderstanding how the tests are being run. They are run inside the VM :)

We are using `vmtest/gotest` which is using a special init process to run the built go binary test inside the VM and output a JSON format that the library is passing forward as a test.

This is why you are including `guest.SkipIfNotInVM(t)` with the test, to check if we are being run inside the qemu VM.

It's quite clever and I recommend reading over the code in `vmtest`.

https://github.com/hugelgupf/vmtest/blob/main/govmtest/gotest.go

Foxboron replied:
As an example, this if how `go-uefi` is capable of doing fairly simple tests.

https://github.com/Foxboron/go-uefi/blob/master/tests/tests/secure_boot_enabled/secure_boot_enabled_test.go

jimmykarily replied:
Oh true :facepalm: ! I just can't get used to it yet. My mind keep gravitating to the types of tests I'm used to :D. 
It makes sense now. Let me change the tests.

jimmykarily replied:
I tried something. I pushed it as a separate commit for easier review. We can squash later. Let me know what you think.

Foxboron replied:
Thanks! Looks much better now :)

Foxboron commented on tests/integrations/export_enrolled_keys/export_enrolled_keys_test.go:
```diff
@@ -39,10 +47,36 @@ func TestExportEnrolledKeysEsl(t *testing.T) {
+}
+
+func findFileByPattern(dirPath string, pattern string) (string, error) {
+	files, err := ioutil.ReadDir(dirPath)
```
`os.ReadDir`, the `ioutil` library is deprecated.

jimmykarily replied:
good catch! My vim-go/coc setup is completely dysfunctional by the build constraints at the top. I haven't found a way to fix that yet. Anyway, fixed.

Foxboron commented on tests/integrations/export_enrolled_keys/export_enrolled_keys_test.go:
```diff
@@ -21,14 +27,16 @@ func TestExportEnrolledKeysDer(t *testing.T) {
-	out, err = utils.ExecWithOutput(fmt.Sprintf("openssl x509 -in %s -inform DER -text -noout", platform))
-	g.Expect(err).ToNot(HaveOccurred(), out)
-	g.Expect(out).To(MatchRegexp("Issuer: C=Platform Key, CN=Platform Key"))
+	derBytes, err := ioutil.ReadFile(platformKey)
```
`os.ReadFile`

jimmykarily commented:
Squashed commits

Foxboron commented:
Thanks!
